### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by regislebrun**
This PR provides several small improvements/bugfixes:
* BernsteinCopulaFactory: when the ResourceMap keys "BernsteinCopulaFactory-MaxM" and "BernsteinCopulaFactory-MinM" were equal, there was a division by zero making OT crash.
* TrapezoidalFactory: now the constraints on the parameters are enforced using a tolerance adapted to the optimization algorithm capabilities
* Lilliefors test: it improves FittingTest::Kolmogorov when called with a distribution factory. Now the pvalue is computed used both a bound on the standard deviation of the Monte Carlo estimate, a minimum and a maximum number of iterations. And now it has the correct name.
* Add 0 to most of the low discrepancy sequences, as suggested in issue  #1601. It is only a first step toward the resolution of this issue.
* Improved Fehlberg, now the step size can be bounded from below, to avoid the algorithm being stucked on stiff problems.
* Minor improvement in ConditionedGaussianProcess: now we use the vectorized version of SpecFunc.rNormal()

### CODE REVIEW

**Review by jschueller — APPROVED** [review]
+1
just a citation error (Lilliefors1967) and this is good to go

**Comment by jschueller** [comment]
oh, you forgot to add the rst entries for lillifors, and and a changelog entry as it deprecates one of the kolmogorov variant

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 1]
```diff
@@ -259,6 +259,15 @@ Distribution FittingTest::BestModelBIC(const Sample  & sample,
 Distribution FittingTest::BestModelKolmogorov(const Sample & sample,
     const DistributionFactoryCollection & factoryCollection,
     TestResult & bestResult)
+{
+  LOGWARN("FittingTest::BestModelKolmogorov(const Sample & sample, const DistributionFactoryCollection & factoryCollection, TestResult & bestResult) is deprecated, use FittingTest::BestModelLilliefors(const Sample & sample, const DistributionFactoryCollection & factoryCollection, TestResult & bestResult)");
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx:263]: shorter version:
"FittingTest::BestModelKolmogorov(Sample, DistributionFactoryCollection, TestResult) is deprecated, use BestModelLilliefors"

File: lib/src/Uncertainty/StatTests/FittingTest.cxx [hunk 2]
```diff
@@ -483,27 +493,62 @@ TestResult FittingTest::Kolmogorov(const Sample & sample,
                                    Distribution & estimatedDistribution,
                                    const Scalar level)
 
+{
+  LOGWARN("FittingTest::Kolmogorov(const Sample & sample, const DistributionFactory & factory, Distribution & estimatedDistribution, const Scalar level) is deprecated. Use FittingTest::Lilliefors(const Sample & sample, const DistributionFactory & factory, Distribution & estimatedDistribution, const Scalar level)");
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/FittingTest.cxx:497]: here too

File: python/src/FittingTest_doc.i.in [hunk 3]
```diff
@@ -401,54 +398,14 @@ TypeError :
...
-
-- Otherwise, the distribution is estimated using the given factory based 
-  on the given sample and the distribution of the test statistics is estimated 
-  using a Monte Carlo approach. 
-  This algorithm is known as Lilliefors's test [Lilliefors1967]_.
-  The sample size can be configured with the *FittingTest-KolmogorovSamplingSize* 
-  key in :class:`~openturns.ResourceMap`. 
-  In this case, there are two output arguments, which are `fitted_dist` 
-  and `test_result`.
+The distribution is supposed to be fully specified, ie no parameter
```
Comment (josephmure) [inline python/src/FittingTest_doc.i.in:401]:
```suggestion
The distribution is supposed to be fully specified, i.e. no parameter
```

File: python/src/FittingTest_doc.i.in [hunk 4]
```diff
@@ -246,7 +246,7 @@ Parameters
 ----------
 sample : 2-d sequence of float
     Tested sample.
-models : list of :class:`~openturns.Distribution` or :class:`~openturns.DistributionFactory`
+models : list of :class:`~openturns.Distribution`
```
Comment (josephmure) [inline python/src/FittingTest_doc.i.in:249]: Why are Distribution Factories no longer allowed?

**Review by josephmure — CHANGES_REQUESTED** [review]
Thanks for these fixes @regislebrun! I have only a few questions/suggestions.

File: lib/etc/openturns.conf.in [hunk 5]
```diff
@@ -213,19 +214,19 @@
...
-  <HaselgroveSequence-InitialSeed value_int="1" />
+  <HaselgroveSequence-InitialSeed value_int="0" />
 
   <!-- OT::ReverseHaltonSequence parameters -->
-  <ReverseHaltonSequence-InitialSeed value_int="1" />
+  <ReverseHaltonSequence-InitialSeed value_int="0" />
 
   <!-- OT::SobolSequence parameters -->
-  <SobolSequence-InitialSeed value_int="1" />
+  <SobolSequence-InitialSeed value_int="0" />
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/etc/openturns.conf.in:229]: Two questions, two answers:
+ no, I cannot convince the Sobol sequence to start at 0 by setting the initial value of the seed to 0 (any clue on how to fix the Sobol sequence is welcomed!). What you see is just a garbage of my failed attempts to solve it, thanks for catching it!
+ yes, openturns.conf is usefull as it allows to overwrite the default values hard-coded into the C++ by values more suited by an application relying on OT, for example the Anatole tolerancing workbench in CATIA. But this file has to be found by the library at startup in order to be taken into account, otherwise this is the hard-coded values that are taken into account. The rule is to have the exact same values as the hard-coded ones in this file by default, and if the user change them he has to make sure his local installation of OT find the file at startup.

File: python/src/FittingTest_doc.i.in (hunk 4 above)
Comment (regislebrun) [reply python/src/FittingTest_doc.i.in:249]: because now you have a FittingTest::BestModelLilliefors() method to do that ;-)! The previous documentation was documenting two different methods with the same name but different arguments, Now you have two different names for these methods.

**Comment by regislebrun** [comment]
> +1
> just a citation error (Lilliefors1967) and this is good to go

What's wrong with this citation? I kept it from the previous documentation. Is there a formatting error?

**Comment by jschueller** [comment]
you deleted the docstring that used that reference, so sphinx errors

**Comment by jschueller** [comment]
You can use Liliefors instead of the deprecated method in python/doc/examples/py_data_analysis/statistical_hypothesis_testing/plot_kolmogorov_test.py, plot_fitted_distribution_ranking.py examples

**Review by josephmure — APPROVED** [review]

**Comment by josephmure** [comment]
Thanks for the explanations @regislebrun!

File: lib/etc/openturns.conf.in [hunk 6]
```diff
@@ -788,9 +788,11 @@
...
 
   <!-- OT::FittingTest parameters -->
-  <FittingTest-ChiSquaredCheckSample     value_bool="true" />
-  <FittingTest-ChiSquaredMinimumBinCount value_int="5"     />
-  <FittingTest-KolmogorovSamplingSize    value_int="10"    />
+  <FittingTest-ChiSquaredCheckSample         value_bool="true" />
+  <FittingTest-LillieforsPrecision           value_float="5"   />
+  <FittingTest-ChiSquaredMinimumBinCount     value_int="5"     />
+  <FittingTest-LillieforsMinimumSamplingSize value_int="10"    />
+  <FittingTest-LillieforsMaximumSamplingSize value_int="10"    />
```
Comment (jschueller) [inline lib/etc/openturns.conf.in:795]: you set 100000 in ResourceMap.cxx

**Comment by regislebrun** [comment]
@josephmure @mbaudin47 I finally found an ugly way to fix SobolSequence (in fact I can't see another way to do it as 0 is special in the Sobol recurrence) and it works. Ideally we should hard-code the initial value of 0 or at least set it as the default value for the initial seed, but it breaks so many tests I prefer to postpone this change after the release, or at least do it in a separate PR where the scrambling would be implemented too.

**Comment by jschueller** [comment]
@regislebrun now you have a conflict after jm branch was merged

**Comment by mbaudin47** [comment]
For reference, this PR will fix #1217

**Comment by mbaudin47** [comment]
Let me rephrase the proposal for Lilliefors'test. The new `FittingTest::Lilliefors` function aims at estimating accurately enough the p-value based on Monte-Carlo estimate. The goal is to generate the smallest Monte-Carlo sample which allows to estimate the p-value sufficiently accurately. The algorithm stops if the accuracy is lower than a threshold (based on the variance estimate) or if the budget is too short (based on the sample size). Is this a correct description?

In the issue #1061, I suggested to use a shortcut. Denote by p1 the p-value evaluated assuming that the parameters are known. The p-value p1 is fast to compute. Denote p2 the p-value assuming that the parameters are estimated. We always have p2<p1 (because the KS statistics is always lower when the parameters are estimated from the sample, so that Lilliefors - complementary - p-value is always smaller than KS's classical p-value). Therefore, an idea is that, if p1 is lower than, say, 0.001, then we do not have to compute p2 because we can reject the proposal distribution. The issue here is that `TestResult` cannot store an information such as "p-value < 0.001": it stores the p-value itself. This shortcut would further improve the speed, do you agree on this point?

File: lib/etc/openturns.conf.in (hunk 6 above)
Comment (regislebrun) [reply lib/etc/openturns.conf.in]: Oupsss! Fixed

**Comment by regislebrun** [comment]
@mbaudin47 Your proposal is interesting, but IMO it should be put into another function as it is neither a Kolmogorov test nor a Lilliefors test. The best I can imagine is a Lilliefors test with a *fast* flag telling if we agree on a quick rejection based on Kolmogorov p-value. The following figure is interesting as it shows the discrepancy between the distribution of the p-value under the null hypothesis assuming that the statistics follows Kolmogorov's distribution instead of Lilliefors's one:
pvalue_KS_fitted_vs_known_size=10_iterMax=040131_KSSize=100000
The red curve corresponds to Lilliefors's p-value distribution, which is uniform over [0,1], and the blue one corresponds to Kolmogorov's p-value, which is clearly right-skewed. These curves have been generated using 40000 samples of both p-values computed on Normal(5.0, 2.0) samples of size 10. Liliefors's p-value is computed using 10^5 Monte Carlo simulations.

Anyway, even if we propose a fast version of FittingTest::Lilliefors, it is dangerous to use it in a fast  version of FittingTest.BestModelLilliefors as the ranking obtained using Kolmogorov's p-value may be different from the one obtained using Lilliefors's p-value (as far as I know).
